feat(MCP): add Snyk Learn tool [IDE-1213]

rrama opened the pull request:
# 🚀 MCP Learn Tools Integration + Code Quality Improvements

## 📋 Overview
This PR introduces MCP (Model Context Protocol) integration with Snyk Learn tools, along with significant code quality improvements, UI enhancements for ignore functionality, and comprehensive testing additions. The changes span multiple components to provide a cohesive developer experience.

## ✨ Major Features Added

### 🔧 MCP Learn Tools Integration
- **`snyk_open_learn_lesson`** - Opens relevant Snyk Learn lessons in browser based on vulnerability context
- **Smart parameter handling** - Supports CVEs, CWEs, rules, ecosystems, and issue types  
- **Ecosystem mapping** - Automatic translation between package managers and learn ecosystems
- **URL enhancement** - Adds IDE context parameters for better lesson targeting
- **Factory pattern implementation** - Clean dependency injection for learn service

### 📋 Development Experience Enhancements
- **Cursor Rules file** - Complete development guidelines for the project
- **Type safety improvements** - Added `OpenBrowserFunc` type definition

## 🐛 Critical Bug Fixes

### Learn Service Improvements
- **Fixed `updateCaches()` return value** - Now returns filtered lessons instead of `nil`
- **Enhanced lesson filtering** - Removes unsupported lessons based on description text
- **Resolved empty lesson list bug** - `GetAllLessons()` now returns actual results
- **Simplified ecosystem handling** - Removed redundant npm->javascript mapping

### Code Quality & Testing
- **81.9% test coverage** across MCP extension package
- **560+ lines of comprehensive tests** - Edge cases, error scenarios, mock integration
- **Linting compliance** - Fixed HTTP method constants, empty branches, unused variables
- **Reduced cyclomatic complexity** - Extracted helper functions for maintainability

### Checklist

- [x] Tests added and all succeed
- [ ] Regenerated mocks, etc. (`make generate`)
 - N/A
- [x] Linted (`make lint-fix`)
- [x] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced
 - N/A


[IDE-1213]: https://snyksec.atlassian.net/browse/IDE-1213?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rrama committed:
feat(MCP): add Snyk Learn tools

One tool for getting all available lessons and one for opening a lesson in the browser.

bastiandoetsch commented on mcp_extension/learn_tool.go:
```diff
@@ -0,0 +1,174 @@
+			return nil, err
+		}
+
+		logger.Info().Str("lessonTitle", lessonTitle).Msg("Parsed parameters for snyk_open_learn_lesson")
```
does this need to be an info log? why?

bastiandoetsch commented on mcp_extension/learn_tool.go:
```diff
@@ -0,0 +1,174 @@
+			}
+		}
+
+		logger.Info().Str("lessonURL", lessonURL).Msg("Attempting to open lesson URL in browser")
```
this shouldn't be info level

bastiandoetsch commented on mcp_extension/learn_tool_test.go:
```diff
@@ -0,0 +1,282 @@
+
+import (
+	"testing"
+	// Core Go testing package
```
well, these comments are really useless. AI generated?

bastiandoetsch commented on mcp_extension/learn_tool_test.go:
```diff
@@ -0,0 +1,282 @@
+)
+
+// learnTestFixture defines the test fixture.
+// learnTestFixture defines the test fixture.
```
duplicate line

bastiandoetsch commented on mcp_extension/learn_tool_test.go:
```diff
@@ -0,0 +1,282 @@
+		return mockLearnSvc
+	}
+
+	// Initialize McpLLMBinding (as done in scan_tool_test.go for its fixture)
+	// We don't need a real CLI path for these tests as learn tools don't directly call CLI.
```
the comments are a bit weird

basti-snyk committed:
feat: add MCP learn tool functionality with comprehensive tests [IDE-1213]

- Add learn tool handlers for getting all lessons and opening specific lessons
- Implement snykGetAllLearnLessonsHandler to retrieve and return lesson data as JSON
- Implement snykOpenLearnLessonHandler with support for SCA and SAST issue types
- Add URL parameter injection (loc=ide) and proper browser opening functionality
- Handle comma-separated CVEs/CWEs and missing optional parameters gracefully
- Create comprehensive unit tests with 100% coverage for all handlers
- Add cursor rules file converted from windsurf rules for consistent development
- Fix JSON linting error in snyk_tools.json (trailing comma)
- Use safe type assertions to prevent runtime panics
- Eliminate code duplication with helper functions in tests

All tests pass (13/13) and code follows project standards.

basti-snyk committed:
refactor: remove obsolete LessonOutput tests [IDE-1213]

Remove TestLessonOutput function and its 2 test cases since the
LessonOutput struct has been deleted from the main code.

This cleanup eliminates unnecessary tests for marshaling/unmarshaling
functionality that no longer exists, maintaining clean test coverage
for only the essential learn tool functionality.

Tests status: 11/11 learn tests passing, 71/71 total mcp_extension tests passing

basti-snyk committed:
refactor: remove snykGetAllLearnLessonsHandler functionality [IDE-1213]

- Remove snykGetAllLearnLessonsHandler function from learn_tool.go
- Remove all associated test cases from learn_tool_test.go
- Remove handler registration from scan_tool.go
- Remove unused LessonOutput struct and JSON marshaling logic
- Clean up unused encoding/json imports

This removes the ability to get all lessons as JSON, streamlining the
learn tool to focus on the lesson opening functionality only.

bastiandoetsch commented:
/gemini review

basti-snyk committed:
Merge branch 'main' into feat/IDE-1213_MCP-learn-tools

basti-snyk committed:
fix: resolve nil pointer issues and improve code quality

- Fixed updateCaches() to return filtered lessons instead of nil
- Added nil check for targetLesson before accessing properties
- Reduced cyclomatic complexity by extracting helper functions
- All tests passing, linting clean, security scans show no issues

[IDE-1213]

bastiandoetsch committed:
docs: updated licenses

basti-snyk committed:
test: improve MCP extension test coverage to 81.9%

- Add comprehensive tests for middleware, shutdown, and server handling
- Test nil lesson handling for recent bug fix
- Add tests for Started status checking and HTTP security validation
- Improve coverage for core learn tool functionality
- All critical security functions now have 100% test coverage

Coverage improvements:
- middleware: 0% → 100%
- Started: 0% → 100%
- snykOpenLearnLessonHandler: 88.9% → 100%
- Shutdown: 50% → 83.3%
- Overall: 78.6% → 81.9%

[IDE-1213]

basti-snyk committed:
fix: resolve linting issues in MCP extension tests

- Fix empty branch in recovery function by explicitly ignoring recovered value
- Replace string literals with http.MethodGet constants for HTTP methods
- Maintain 81.9% test coverage with all tests passing
- Code now adheres to all Go linting standards and stdlib conventions

[IDE-1213]

basti-snyk committed:
refactor: remove remaining references to deleted getAllLearnLessons functionality

- Remove SnykGetAllLearnLessons constant from scan_tool.go
- Update README.md to remove snyk_get_all_learn_lessons from tool list
- Complete cleanup of previously removed getAllLearnLessons handler
- Maintain consistency between code and documentation

[IDE-1213]

basti-snyk committed:
style: fix constant alignment formatting in scan_tool.go

- Align constant declarations with consistent spacing
- Resolve goimports linting violation on line 35
- Maintain backward compatibility and functionality
- All tests passing, no logic changes

[IDE-1213]

basti-snyk committed:
Merge branch 'main' into feat/IDE-1213_MCP-learn-tools

ShawkyZ commented on mcp_extension/learn_tool.go:
```diff
@@ -0,0 +1,135 @@
+		issueTypeString := parseStringParam(args, "issueType")
+		issueType := parseIssueType(issueTypeString)
+
+		targetLesson, err := m.learnService.GetLesson(ecosystemString, ruleString, cweArray, cveArray, issueType)
```
check if  m.learnService != nil

ShawkyZ commented:
Should we wait on merging this change until the custom output PR is implemented ?

bastiandoetsch commented on mcp_extension/learn_tool.go:
```diff
@@ -0,0 +1,135 @@
+		return "", fmt.Errorf("invalid lesson URL: %w", err)
+	}
+	q := u.Query()
+	q.Set("loc", "ide")
```
I've reached out to the team to verify.

basti-snyk committed:
chore: update ecosystem param description

basti-snyk committed:
refactor: use global logger for browser stdout

basti-snyk committed:
fix: add nil check, don't return errors but result texts

basti-snyk committed:
fix: use mcp as url param

basti-snyk committed:
fix: tests
